examples/Cassie/cassie_rbt_state_estimator.cc

```diff
@@ -1,140 +1,981 @@
+      output.GetVelocities(), true);
+
+  // M, C and B matrices
+  MatrixXd M = tree_.massMatrix(cache);
+  const typename RigidBodyTree<double>::BodyToWrenchMap no_external_wrenches;
+  MatrixXd C = tree_.dynamicsBiasTerm(cache, no_external_wrenches);
+  MatrixXd B = tree_.B;
+
+  // control input - obtained from cassie_input temporarily
+  VectorXd u = output.GetEfforts();
+
+  // Jb - Jacobian for fourbar linkage
+  MatrixXd Jb = tree_.positionConstraintsJacobian(cache, false);
+  VectorXd Jb_dot_times_v = tree_.positionConstraintsJacDotTimesV(cache);
+
+  /** Jc{l, r}{f, r} - contact Jacobians **/
+  // l - left; r - right; f - front; r - rear
+  MatrixXd Jclf = tree_.transformPointsJacobian(cache,
+      front_contact_disp_, left_toe_ind, 0, false);
+  MatrixXd Jclr = tree_.transformPointsJacobian(cache,
+      rear_contact_disp_, left_toe_ind, 0, false);
+
+  MatrixXd Jcl(3*2, Jclf.cols());
+  Jcl.block(0, 0, 3, Jclf.cols()) = Jclf;
+  Jcl.block(3, 0, 3, Jclr.cols()) = Jclr;
+
+  MatrixXd Jcrf = tree_.transformPointsJacobian(cache,
+      front_contact_disp_, right_toe_ind, 0, false);
+  MatrixXd Jcrr = tree_.transformPointsJacobian(cache,
+      rear_contact_disp_, right_toe_ind, 0, false);
+
+  MatrixXd Jcr(3*2, Jcrf.cols());
+  Jcr.block(0, 0, 3, Jcrf.cols()) = Jcrf;
+  Jcr.block(3, 0, 3, Jcrr.cols()) = Jcrr;
+
+  // Contact jacobian dot times v
+  VectorXd Jclf_dot_times_v = tree_.transformPointsJacobianDotTimesV(
+      cache, front_contact_disp_, left_toe_ind, 0);
+  VectorXd Jclr_dot_times_v = tree_.transformPointsJacobianDotTimesV(
+      cache, rear_contact_disp_, left_toe_ind, 0);
+
+  VectorXd Jcl_dot_times_v(6);
+  Jcl_dot_times_v.head(3) = Jclf_dot_times_v;
+  Jcl_dot_times_v.tail(3) = Jclr_dot_times_v;
+
+  VectorXd Jcrf_dot_times_v = tree_.transformPointsJacobianDotTimesV(
+      cache, front_contact_disp_, right_toe_ind, 0);
+  VectorXd Jcrr_dot_times_v = tree_.transformPointsJacobianDotTimesV(
+      cache, rear_contact_disp_, right_toe_ind, 0);
+
+  VectorXd Jcr_dot_times_v(6);
+  Jcr_dot_times_v.head(3) = Jcrf_dot_times_v;
+  Jcr_dot_times_v.tail(3) = Jcrr_dot_times_v;
+
+  // double left_leg_velocity = (Jcl*output.GetVelocities()).norm();
+  // double right_leg_velocity = (Jcr*output.GetVelocities()).norm();
+
+  MatrixXd J_imu = tree_.transformPointsJacobian(cache,
+      imu_pos_, pelvis_ind, 0, false);
+  VectorXd J_imu_dot_times_v = tree_.transformPointsJacobianDotTimesV(
+      cache, imu_pos_, pelvis_ind, 0);
+
+  Vector3d alpha_imu = output.GetIMUAccelerations();
+
+  RigidBody<double>* pelvis_body = tree_.FindBody("pelvis");
+  Isometry3d pelvis_pose = tree_.CalcBodyPoseInWorldFrame(cache,
+      *pelvis_body);
+  MatrixXd R_WB = pelvis_pose.linear();
+  Vector3d gravity_in_pelvis_frame = R_WB.transpose()*gravity_;
+  alpha_imu -= gravity_in_pelvis_frame;
+
+  std::vector<double> optimal_cost;
+
+  // Mathematical program - double contact
+  MathematicalProgram quadprog_double;
```

Comment:
This QP should be created once, not inside the loop. In the loop, just update it and solve. This is probably something we can leave, provided someone creates an issue to resolve it soon.

Reply:
Created an issue that aims to resolve this. #76